components/table/Table.tsx

```diff
@@ -148,89 +270,38 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
         'fixed columns instead, see: https://u.ant.design/fixed-columns.',
     );
 
-    if (expandedRowRender && columns.some(({ fixed }) => !!fixed)) {
+    if (expandedRowRender && columnsProp.some(({ fixed }) => !!fixed)) {
       warning(
         false,
         'Table',
         '`expandedRowRender` and `Column.fixed` are not compatible. Please use one of them at one time.',
       );
     }
 
-    this.columns = props.columns || normalizeColumns(props.children as React.ReactChildren);
-
-    this.createComponents(props.components);
+    const columns = columnsProp || normalizeColumns(props.children as React.ReactChildren);
 
     this.state = {
-      ...this.getDefaultSortOrder(this.columns),
+      ...this.getDefaultSortOrder(columns),
       // 减少状态
-      filters: this.getFiltersFromColumns(),
+      filters: getFiltersFromColumns<T>(),
       pagination: this.getDefaultPagination(props),
       pivot: undefined,
+      prevProps: props,
+      components: createComponents(props.components, undefined, true),
+      columns,
     };
-
-    this.CheckboxPropsCache = {};
-
-    this.store = createStore({
-      selectedRowKeys: getRowSelection(props).selectedRowKeys || [],
-      selectionDirty: false,
-    });
   }
 
-  componentWillReceiveProps(nextProps: TableProps<T>) {
-    this.columns = nextProps.columns || normalizeColumns(nextProps.children as React.ReactChildren);
-    if ('pagination' in nextProps || 'pagination' in this.props) {
-      this.setState(previousState => {
-        const newPagination = {
-          ...defaultPagination,
-          ...previousState.pagination,
-          ...nextProps.pagination,
-        };
-        newPagination.current = newPagination.current || 1;
-        newPagination.pageSize = newPagination.pageSize || 10;
-        return { pagination: nextProps.pagination !== false ? newPagination : emptyObject };
-      });
-    }
-    if (nextProps.rowSelection && 'selectedRowKeys' in nextProps.rowSelection) {
-      this.store.setState({
-        selectedRowKeys: nextProps.rowSelection.selectedRowKeys || [],
-      });
-    } else if (this.props.rowSelection && !nextProps.rowSelection) {
-      this.store.setState({
-        selectedRowKeys: [],
-      });
-    }
-    if ('dataSource' in nextProps && nextProps.dataSource !== this.props.dataSource) {
-      this.store.setState({
-        selectionDirty: false,
-      });
-    }
-
-    // https://github.com/ant-design/ant-design/issues/10133
-    this.CheckboxPropsCache = {};
-
-    if (this.getSortOrderColumns(this.columns).length > 0) {
-      const sortState = this.getSortStateFromColumns(this.columns);
+  componentDidUpdate() {
```

Comment:
componentDidUpdate lifecycle maybe not need. getDerivedStateFromProps do same thing

Reply:
Maybe `getSnapshotBeforeUpdate` is more suitable